### PR TITLE
Rescheduler design space doc.

### PR DESCRIPTION
**Pull request by davidopp**
@erictune @bgrant0607 @timothysc

Ref #12140

### CODE REVIEW

**Comment by timothysc** [comment]
cc @smarterclayton @eparis

File: docs/proposals/rescheduler.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+defined by some particular label selector, a maximum number of Pods
+that can be down at any one time among a set defined by some
+particular label selector, etc. These policies are presumably part of
+the Rescheduler's configuration.
+
+There are a lot of design possibilities for a rescheduler. To explain
+them, it's easiest to start with the description of a baseline
+rescheduler, and then describe possible modifications. The Baseline
+rescheduler
+* only kicks in when there are one or more PENDING Pods for some period of time; its objective function is binary: completely happy if there are no PENDING Pods, and completely unhappy if there are PENDING Pods; it does not try to optimize for any other aspect of cluster layout
```
Comment (timothysc) [inline docs/proposals/rescheduler.md]: Aren't these policy choices, and thus more recommendations vs. hard and fast rules?

File: docs/proposals/rescheduler.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+Because rescheduling is disruptive--it causes one or more
+already-running Pods to die when they otherwise wouldn't--a key
+constraint on rescheduling is that it must be done subject to
+disruption SLOs. There are a number of ways to specify these SLOs--a
+global rate limit across all Pods, a rate limit across a set of Pods
+defined by some particular label selector, a maximum number of Pods
+that can be down at any one time among a set defined by some
+particular label selector, etc. These policies are presumably part of
+the Rescheduler's configuration.
+
```
Comment (timothysc) [inline docs/proposals/rescheduler.md]: I really like the way the doc reads, but there is a key use-case which I find that is missing.  

Gracefully shutting down underutilized nodes, which could be driven by policy.

File: docs/proposals/rescheduler.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+* If it is going to run continuously in the background to optimize cluster layout and is going to act as a scheduler rather than just killing Pods, then it needs to know the predicate functions and some compatible (but not necessarily identical) priority functions  One example of a case where "compatible but not identical" might be useful is if the main scheduler(s) has a very simple scheduling policy optimized for low scheduling latency, and the Rescheduler having a more sophisticated/optimal scheduling policy that requires more computation time. The main thing to avoid is for the scheduler(s) and rescheduler to have incompatible priority functions, as this will cause them to "fight" (though it still can't lead to an infinite loop, since the scheduler(s) only ever touches a Pod once).
+
+The vast majority of users probably only care about rescheduling for three scenarios:
+
+1. Redistribute Pods onto new nodes added by a cluster auto-scaler
+1. Move Pods around to get a PENDING Pod to schedule
+1. Move Pods around when CPU starvation is detected on a node
```
Comment (timothysc) [inline docs/proposals/rescheduler.md]: I would start the document with use cases, as it drives motivation.  

+ the use case I mentioned above.

**Comment by timothysc** [comment]
Good stuff, only a couple of comments.

File: docs/proposals/rescheduler.md (hunk 2 above)
Comment (davidopp) [inline docs/proposals/rescheduler.md]: I tried to cover this use case in the last sentence of the doc. Is there additional description you would suggest there? (Or is that actually not the use case you had in mind?)

File: docs/proposals/rescheduler.md (hunk 1 above)
Comment (davidopp) [inline docs/proposals/rescheduler.md]: Yes. I considered writing the doc in a more traditional style where I lay out the axes of the design space and explain the different points in the space. However, after struggling with that a bit, I realized it was easier to start with one point in the design space (the Baseline rescheduler) and then describe variations on it. But either way, this is just talking about different design possibilities. Let me know if I am not fully understanding your point.

**Comment by bgrant0607** [comment]
Given that this is non-urgent, I probably won't have time to review it.

File: docs/proposals/rescheduler.md (hunk 3 above)
Comment (davidopp) [inline docs/proposals/rescheduler.md]: OK I'll add a section at the beginning calling out the use cases.

**Comment by davidopp** [comment]
@timothysc OK I implemented your suggestions to call out the use cases explicitly and to make the "gracefully shutting down underutilized nodes" use case clearer. Let me know what you think.

**Comment by timothysc** [comment]
LGTM 

+ I believe this could also be used to address openshifts idling 
cc @derekwaynecarr @ncdc

**Comment by erictune** [comment]
Thanks for writing this David: thorough treatment of the topic.

Nits on the text: 
 - define term SLO before use 
- fix formatting of numbered list in appendix.

**Comment by erictune** [comment]
What about a thing that only detects when you need to add a node, and only removes completely empty nodes?  That would allow us to demonstrate horizontal autoscaling, and would possibly be simpler, because it would not require defining an SLO and would not attempt bin-packing optimizations.

I think horizontal autoscaling might be pretty compelling for customers by itself, more so than improved bin-packing.  Bin packing might save, say 10% of resources.  The value of saving, say 3 x n1-standard-1 instances out of a cluster of say, 30 nodes, is a dollar or two an hour.  Having an skilled ops person on call 24x7 to handle a possible traffic spike costs a lot more than a few dollars an hour.

**Comment by erictune** [comment]
lgtm.   fine for comments to be addressed later.

**Comment by derekwaynecarr** [comment]
Thanks for the detailed write-up.  This was really useful in framing future
discussion.

**Pull request merged**